editor.planx.uk/src/lib/local.idb.ts

```diff
@@ -0,0 +1,161 @@
+        throw Error(
+          "Migration from localStorage failed - not upgrading IndexedDB",
+        );
+      }
+    },
+    // we include logs to debug cases where multiple db connections conflict (unlikely to occur under v1))
+    blocked() {
+      console.warn(
+        "IndexedDB upgrade is blocked for this tab, likely because another tab has an open connection",
+      );
+    },
+    blocking() {
+      console.warn(
+        "IndexedDB connection on this tab is blocking another tab from upgrading the database",
+      );
+    },
+  });
+};
+
+export const getLocalFlowIdb = async (
+  flowId: string,
+): Promise<Session | undefined> => {
+  console.debug(
+    `Retrieving flow with ID ${flowId} from IndexedDB (if it exists)`,
+  );
+  let db: IDBPDatabase<PlanXIDB> | undefined;
+  try {
+    db = await getDb();
+    const session = await db.get(DB_STORE_FLOW_SESSIONS, flowId);
+    return session;
+  } catch (e) {
+    console.error(`Error retrieving flow with ID ${flowId} from IndexedDB:`, e);
+    return undefined;
+  } finally {
+    // in any case, we close the db connection to avoid memory leaks
+    db?.close();
+  }
+};
+
+export const setLocalFlowIdb = async (
+  flowId: string,
+  session: Session,
+): Promise<void> => {
+  console.debug(`Storing flow with ID ${flowId} to IndexedDB`);
+  let db: IDBPDatabase<PlanXIDB> | undefined;
+  try {
+    db = await getDb();
+    await db.put(DB_STORE_FLOW_SESSIONS, session, flowId);
+  } catch (e) {
+    console.error(`Error storing flow with ID ${flowId} in IndexedDB:`, e);
+  } finally {
+    db?.close();
+  }
+};
+
+export const clearLocalFlowIdb = async (flowId: string): Promise<void> => {
+  console.debug(`Clearing flow with ID ${flowId} from IndexedDB`);
+  let db: IDBPDatabase<PlanXIDB> | undefined;
+  try {
+    db = await getDb();
+    await db.delete(DB_STORE_FLOW_SESSIONS, flowId);
+  } catch (e) {
+    console.error(`Error clearing flow with ID ${flowId} from IndexedDB:`, e);
+  } finally {
+    db?.close();
+  }
+};
+
+// we need a method for migrating existing sessions across from localStorage
+export const migrateLocalStorageToIdb = async (
+  tx: IDBPTransaction<PlanXIDB, StoreNames<PlanXIDB>[], "versionchange">,
+): Promise<number> => {
+  console.debug(`Checking for flows in localStorage to migrate to IndexedDB`);
+  try {
+    // get all flow IDs present in localStorage
+    const keys = Object.keys(localStorage);
+    const flowIds = keys
+      .filter((key) => key.startsWith(LOCAL_STORAGE_KEY_PREFIX))
+      .map((key) => key.replace(LOCAL_STORAGE_KEY_PREFIX, ""));
+
+    // exit if no flows to migrate
+    if (flowIds.length === 0) {
+      console.debug("No flows in localStorage to migrate");
+      return 0;
+    }
+
+    console.log(
+      `Migrating ${flowIds.length} flow(s) from localStorage to IndexedDB`,
+    );
+    const store = tx.objectStore(DB_STORE_FLOW_SESSIONS);
+    const failed_migrations = new Set<string>();
```

Comment:
nit: 
```suggestion
    const failedMigrations = new Set<string>();
```

🐍  😉

Reply:
hehe nice catch, that's all the py I've been writing for the [application project](https://mnist.commune.london/) I did coming thru ;)